### ISSUE TITLE
Concatenate layer does not support 'axis' parameter

### ISSUE THREAD

**Issue by krworks** [Dobiasd/frugally-deep#62]
I thought I would point out a bit of an issue I've ran into in regards to Concatenation layer:
Concatenate layer does not support 'axis' parameter as seen on page: https://keras.io/layers/merge/

Keras Documentation says:
> [Usage]
> It takes as input a list of tensors, all of the same shape except for the concatenation axis, and returns a single tensor, the concatenation of all inputs.
> 
> [Arguments]
> axis: Axis along which to concatenate.

The concatenate code in frugally-deep seems to do a width / height sans depth check, the component of depth I'm not sure of in terms of axis number.  Effectively there's nothing passed through from the layer which has no parameters.
```
inline tensor3 concatenate_tensor3s(const tensor3s& ts)
{
    assertion(fplus::all_the_same_on(
        fplus_c_mem_fn_t(tensor3, size_without_depth, shape2), ts),
        "all tensors must have the same width and height");
```

**Comment by Dobiasd** [comment]
Good finding. If we don't support it `convert_model.py` should at least raise an exception in case of `axis!=-1`.

Do you need support for different axis. If do I could implement it. Or if you like to, you could also give it a try in a pull request. :slightly_smiling_face:

**Comment by krworks** [comment]
You are likely best at implementing the code for FunctionalPlus as I'm more C oriented myself.  I'm sure I could do it but it may not be your standards.  If there's no commit tonight/tomorrow then I'll start scratching something together.  If it works then I'll organize it up in the right place (probably belongs in FunctionalPlus, right?) and do a PR

**Comment by krworks** [comment]
Here's a C implementation that seems to give the correct outputs for a variable set of Tensors.
https://gist.github.com/krworks/ce9085c447b3667960d0b8fa72d6feec

**Comment by krworks** [comment]
Oh, I couldn't find in the Keras docs but I assume axis == -1 means last axis

**Comment by Dobiasd** [comment]
Yes, `-1` usually means last axis. I guess this is borrowed from indexing a sequence in Python:
```python
assert [0,1,2,3,4,5][-1] == 5
```

**Comment by Dobiasd** [comment]
This was fun to implement, because of the spatial thinking involved. :)

**Comment by krworks** [comment]
<sarcasm> Leave it to the germans to get great thrills from axis aligned matrix concatenation
